### DISCUSSION TITLE
rpm packaging

### DISCUSSION THREAD

**Discussion by guiv42** [helge17/tuxguitar#13] [General]
Hi,
I am currently working on an implementation for rpm packaging (just like one already exists for deb), taking benefit from the current rework of the build process (thanks!). Yet the latest available version of TG in the rpm world is really outdated (1.5.4).
And this raises a few questions

Question 1: what is the best packaging strategy?
---
I could identify these 2 options:
1. strategy used for deb packaging in this repo: install *everything* in `/opt/tuxguitar` and symlink to items in this folder (start script, .desktop file, etc)
2. strategy used to build the 1.5.4 rpm package (see [here](https://build.opensuse.org/package/view_file/openSUSE:Factory/tuxguitar/tuxguitar.spec?expand=1)): distribute tux-guitar specific files across the filesystem (`/usr/share/tuxguitar`, `/usr/share/java/tuxguitar`, `/usr/lib64/tuxguitar`, `/usr/lib64/java/tuxguitar`, etc.), and share the common resources (swt, itext, default sound font ...)

I have manually produced rpms with both strategies, and I'm currently working on automation.
Looking at the philosophy of packaging in my OpenSUSE environment, I would say strategy 2 seems the *cleanest*. The resulting package is <6Mo, against >60Mo with the first strategy.
But... I must admit this creates an issue that up to now I could not solve. Distributing all resources (.so, .jar, etc) everywhere implies a complete rework of the starting script `tuxguitar.sh`, especially to define correctly the CLASSPATH parameter. Config files `tuxguitar-fluidsynth.cfg` and `tuxguitar-jsa.cfg` would also need to be modified (default sound font).

What were the motivations for this packaging strategy 1. used in Debian environment?

Question 2: what is "VST_SDK"?
---
The official TG1.5.4 rpm package in OpenSUSE has [patched](https://build.opensuse.org/package/view_file/openSUSE:Factory/tuxguitar/no-vst.patch?expand=1) TG pom file to remove the "unavailable VST SDK". From the [build script example](https://github.com/helge17/tuxguitar/blob/master/misc/build_tuxguitar_from_source.sh) here in this repo, I coud find the needed files [here](https://github.com/R-Tur/VST_SDK_2.4), but I must admit this puzzles me:
- readme.md just mentions "Moved to: https://github.com/Iunusov/VST_SDK_2.4"
- this URL leads to a 404 error
- Looking at the available github repo, there is no reference to any license

According to the [rpm package history](https://build.opensuse.org/request/show/821257), VST_SDK has been removed from rpm of 1.5.4 since July 2020 (I did not check the older versions). So at least in my environment it doesn't seem to be required.
The *easiest* path to create a rpm package seems to apply the same patch to pom files during the packaging process (i.e. remove tuxguitar-synth-vst-linux-x86_64). However, it might not be the *cleanest*?
What is exactly this VST_SDK needed for? Would it be possible to remove this dependency? Or to link TG to a more appropriate implementation (if such exists)?

Question 3: what would be the best environment?
---
On my side, I can not build all configurations: I have no access to MacOS or Windows machines, and yet I haven't set up any freeBSD or Debian environment. Not yet configured for Android also. But on the other side I don't know how to provide an easy path for rpm production. For sure I can produce rpms in my environment, but it would probably not be easy in a Debian env.
I'm currently working on pom files to automatize rpm production: maven calls the rpmbuild utility to create the package. But it could be done the other way round: rpmbuild can build from source, calling maven (as it is done with current 1.5.4 rpm).
Any idea to make it simple?

Note: analysis is still ongoing for the other patches currently applied by the 1.5.4 rpm package, other questions might follow!

**Comment by helge17** [comment]
## 1. packaging strategy

I think you have to distinguish between "foreign" packages and packages that come directly from the distribution's repository.

"Foreign" packages don't necessarily follow the distribution's file hierarchy, license or other rules. They are OPTional and are located in /opt. For example, on my system, in addition to TuxGuitar, I find Signal Messenger, Temaviewer and Zoom under /opt. The TuxGuitar packages you find here (or on Sourceforge) contain non-free software, a soundfont with an unclear license, have their own libraries, lack many files that a normal Debian (and RPM) package should have, etc. So I think they are well positioned in /opt.

Native packages from Debian, Fedora, etc. repositories follow the rules of the distribution: placement of files, use of OS libraries as much as possible, free licenses only, etc. Just look at the Debian package https://packages.debian.org/sid/tuxguitar: It comes without the soundfont MagicSFver2.sf2 (~70MB in size, this is why the distribution packages are so much smaller) and the VST plugin. SWT, itextpdf and some other things do not come from the TuxGuitar sources. Instead, the Debian packages are used.

Long story short, I would just keep the files under /opt in this repo.

## 2. VST_SDK

This plugin should make it possible to use "Steinberg Virtual Instruments" in TuxGuitar (https://www.steinberg.net/vst-instruments/). I don't know if anyone still uses it. v2.4 is outdated and not open source, so it can't be used in Debian, Fedora etc packages. See also https://www.steinberg.net/developers/.

As long as it works, I thought I'd keep it, but maybe it's better to remove it, I don't know.

## 3. build environment

Linux, Windows and Android packages are built on Linux, no Windows/Android device needed.

For macOS I use a Hackintosh x86_64 VM, for FreeBSD I use a FreeBSD VM. Btw TuxGuitar can also be built on arm64, ppc64 and other Linux architectures, but this requires the right hardware or a Qemu VM. I did it once on arm64 and ppc64el in Qemu, but it's too much work to keep all these VMs up to date.

Since TuxGuitar is 95% Java, it should be possible to build it on only one platform and use it everywhere (https://en.wikipedia.org/wiki/Write_once,_run_anywhere). But at least the native modules are architecture dependent, so I don't know if this will be possible in the near future....

Although I don't have a RPM system, I will try to build RPM packages e.g. with "alien --to-rpm tuxguitar.deb" on Debian without any new pom.xml and upload them here, just give me some time...

**Comment by guiv42** [comment]
Thank you very much for this very detailed answer. So I think I will try to follow the 2 packaging approaches in parallel:
- build a distribution-specific package, through the official OpenSuse Build Service, from the official TG1.5.6 release and submit it to OpenSuse. Following exactly the same logic as the existing 1.5.4 (patching tuxguitar.sh, vst, soundfount, etc). Basically, just submit a package upgrade.
- for this repo, automatize the production of rpm packages using the same logic as the .deb, keeping everything unchanged (no file patch).

I have had bad experiences in the past with these deb-rpm converters which I do not find very convincing.
I am currently considering the option to build a docker container to make the rpm production easy in any Linux-based environment. The idea would be to copy the output of the existing build chain to a dedicated folder, start the container which then builds the rpm(s). This would mean adding a few lines to the global build script to produce the rpm(s), without any additional pom and without the need for a machine in a specific environment.
Do you think this approach makes sense?
Don't know if I will succeed but I think it is worth a try. Of course if I can get it to work I will make it available.
> just give me some time...

Well, for sure! In fact I am not requesting anything ;-)

**Comment by helge17** [comment]
alien --to-rpm is indeed not the best way to create an RPM file. But since the DEBs from here are quite simple, mainly Java, I wanted to give it a try: The new release 2023-04-19 contains RPMs created with alien.

It's great if you create RPMs for OpenSUSE! I did the same for Debian/Ubuntu. I am not the official maintainer of TuxGuitar, but my version was accepted on https://mentors.debian.net/. For more information see https://tracker.debian.org/pkg/tuxguitar.

> Do you think this approach is useful?

I'm not sure if I understand you correctly: you want to build the binaries with your "existing build chain" and create a Docker container just for building the RPM? Is your build chain not an RPM-based system?
Anyway, if you want to add code to create "native" RPMs, just send a pull request. I might also try building RPM packages on my Debian system. I'm not familiar with creating RPMs, but the commands rpm, rpmbuild, etc. are also available in Debian.

**Comment by guiv42** [comment]
Just received a notification you have published a new release. Were you kidding when you said you needed some time?
I'll test the rpm you published and give you some feedback, thanks.

About OpenSuse, I'm not quite sure it will succeed. Just looking at the descriptive fields of the package, it's not so easy to identify a maintainer, and especially a web site. Without this I don't know if such a package can be accepted. Also the patches to apply to 1.5.6 are not so simple, so I will need some time for that. However I'll have a try, one never knows. Might be useful in the future, depending of what the future of Tuxguitar will be...

About my proposal for a docker container: for sure I can build "native" RPMs in my system, I already have one running from this repo (with everything installed in /opt following our discussion). In fact mine should be very similar to the one you have generated with alien. I'll try to compare.
The idea of a docker container was to give the possibility to other users not using rpm-based system to produce RPMs easily (you, or anyone who would like to do it from your repo). I did not know you had access to rpmbuild in your environment, in this case you may not even need a container.
Let me just a little more time to finetune my rpm packaging, then I can provide the ".spec" file and the associated procedure to produce the rpm. This way you could try it in your environment.

**Comment by guiv42** [comment]
The rpm you released seems to work fine :-)
Work is still in progress on my side to package with rpmbuild. Looking at what I currently have, the diff with the alien-generated version is minor.
I mimicked the files organization defined by your deb pom file (i.e. install in /opt and add symlinks to /opt). Then I added a few specific directives just as they were implemented in TG1.5.4 rpm : man entry, doc, and tuxguitar icons with different resolutions.
The result is:
- rpmbuild compressed the man entry: /usr/share/man/man1/tuxguitar.1.gz
- license file is installed in /usr/share/licenses/tuxguitar/LICENSE
- AUTHORS, README and CHANGES files are installed in /usr/share/doc/packages/tuxguitar/
- icons are stored in /usr/share/icons/hicolor/16x16/apps/tuxguitar.png (same with 32x32, etc)

In the version produced by alien, 2 binaries are symlinked from /usr/lib:
- /usr/lib/.build-id/03/c1e5391c7599d2357509dc793137caa0f5d16e -> ../../../../opt/tuxguitar/vst-client/tuxguitar-synth-vst.bin
- /usr/lib/.build-id/3b/6546682b17a30ca0350125fc19ce022a926e69 -> ../../../../opt/tuxguitar/lv2-client/tuxguitar-synth-lv2.bin
This is a bit more difficult to understand. Anyhow, to be honest I had not yet identified where to store (or symlink from) these 2 binaries.
How can I test if "vst-client" and/or "lv2-client" work correctly?

Pre/post install scripts:
I can find in the alien version the pre/post-install scripts used by the deb package. Guess they are used to refresh the association of file types with the app. I did not add any of these scripts in my rpm, yet it seems to be automatically handled by the package manager. Even without these scripts, clicking a ".tg" file just after install directly opens TuxGuitar.

**Comment by guiv42** [comment]
Update
Current build of a rpm package installing TuxGuitar in /opt works fine, so I suggest to keep this process unchanged (using alien from deb package).
Just for information, I'm currently working on an "official" integration of this version of TuxGuitar in openSUSE. I have successfully built a rpm from 1.6.0 using OpenSUSE Build Service, in a similar way 1.5.4 was built. I just submitted this build.

Edit: request was accepted, TuxGuitar 1.6.0 is in openSUSE :)
